### PR TITLE
ISSUE-1001 Implement backend part of Log Search feature

### LINKED ISSUE #1001: Implement backend part of Log Search feature
The issue is to provide HTTP API which queries to Ambari Infra (Solr) with some criteria, and provide query results to the UI. UI will construct frontend of log search feature based on the API.

This assumes Storm worker log is parsed as pre-defined schema, and published to Ambari Infra. We will take care of this from Ambari side.

### PR DESCRIPTION
**Pull request by HeartSaVioR**
* add AMBARI_INFRA to Ambari services so that it can be imported automatically
* setup interface and container for managing implementations for the TopologyLogSearch interface
* crafted Ambari Infra & Storm pair implementation of TopologyLogSearch
  * new module 'streamline-logsearch-storm' added
* expose LogSearch feature to HTTP API

* [x] Made the API working with secured cluster
* [x] discuss and expose more fields in result

This closes #1001.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
@arunmahadevan @shahsank3t 

1. This patch requires Ambari Infra to be installed to Ambari cluster. Please install it before testing.
2. Importing Ambari cluster will add `AMBARI_INFRA` as one of service: this means you need to reimport again if you would want to use existing cluster. (New icon needed for AMBARI_INFRA.)
3. UI needs to set `logSearchService` to AMBARI_INFRA in namespace if AMBARI_INFRA is selected in the namespace, like what we did for AMBARI_METRICS.
4. To test the patch, we need to push some documents to Ambari Infra manually, and also needs to change schema on existing collection. I'll provide how to do it sooner.

**Comment by HeartSaVioR** [comment]
Below is the capture of dummy document. This is based on actual document and I added some more fields which I expected at the end.

screen shot 2017-11-02 at 5 38 17 pm

And below is the capture of HTTP API result.

screen shot 2017-11-02 at 5 36 19 pm

Please let me know anytime if we want more fields to utilize on UI side.

**Comment by HeartSaVioR** [comment]
Addressed kerberized Ambari cluster as well.

**Comment by HeartSaVioR** [comment]
For making this work with Ambari secured cluster, below configuration should be applied before:

> `streamline_jass.conf` in Streamline:

```
Client {
com.sun.security.auth.module.Krb5LoginModule required
useKeyTab=true
keyTab="<streamline_keytab>"
storeKey=true
useTicketCache=true
debug=true
principal="<streamline_principal>";
};
```

> `security.json` in Ambari Infra:

```
  "authorization": {
    "class": "org.apache.ambari.infra.security.InfraRuleBasedAuthorizationPlugin",
    "user-role": {

...
"<streamline_principal>": ["logsearch_user", ..., "dev"],
...
```

**Comment by HeartSaVioR** [comment]
Also addressed multiple values of log level and component name as well.
@shahsank3t You can use "logLevel" and "componentName" multiple times in query and it will behave as "OR".

**Comment by HeartSaVioR** [comment]
Also change the result of log search API endpoint to include matched doc count.

**Comment by HeartSaVioR** [comment]
Rebased and squashed commits.

**Comment by HeartSaVioR** [comment]
@arunmahadevan Could you review this and also #1037 to merge the pair in?

**Comment by HeartSaVioR** [comment]
Rebased.

**Comment by HeartSaVioR** [comment]
Newer commit addresses the change of dynamic field: using "sdi_*" instead of "storm_*" and "streamline_*". Change from UI side  is not needed, but Storm log config for Ambari log feeder should be changed.

**Comment by HeartSaVioR** [comment]
Rebased.

**Comment by arunmahadevan** [comment]
@HeartSaVioR , a few questions :

> This patch requires Ambari Infra to be installed to Ambari cluster. Please install it before testing.

We need to make sure Ambari auto selects logviewer and infra when SAM is installed. We may need to add the right dependencies in the SAM Ambari definition. Will raise an Ambari issue for this.


> Importing Ambari cluster will add AMBARI_INFRA as one of service: this means you need to reimport again if you would want to use existing cluster. (New icon needed for AMBARI_INFRA.)

How do we plan to handle upgrades ? Can we disable the Log Search feature in the SAM UI until user installs Ambari infra + logsearch and updates their env ?


>UI needs to set logSearchService to AMBARI_INFRA in namespace if AMBARI_INFRA is selected in the namespace, like what we did for AMBARI_METRICS.

Don't get this. Does user need to do anything ? Can you elaborate? 

Other than this, I dont see the log search configuration (grok patterns) as a part of this PR. How do we plan to update Storm's logsearch config when SAM is installed ?

**Comment by HeartSaVioR** [comment]
>> This patch requires Ambari Infra to be installed to Ambari cluster. Please install it before testing.

> We need to make sure Ambari auto selects logviewer and infra when SAM is installed. We may need to add the right dependencies in the SAM Ambari definition. Will raise an Ambari issue for this.

We can do it, but one thing to note that LogSearch feature is now technical preview, not GA. 

>> Importing Ambari cluster will add AMBARI_INFRA as one of service: this means you need to reimport again if you would want to use existing cluster. (New icon needed for AMBARI_INFRA.)

> How do we plan to handle upgrades ? Can we disable the Log Search feature in the SAM UI until user installs Ambari infra + logsearch and updates their env ?

Currently backend responds empty result when Ambari infra is not selected in env. I thought that's acceptable, but we could hide the logsearch button in view mode if env is not having Ambari infra.

>> UI needs to set logSearchService to AMBARI_INFRA in namespace if AMBARI_INFRA is selected in the namespace, like what we did for AMBARI_METRICS.

> Don't get this. Does user need to do anything ? Can you elaborate?

This is for UI side guide and already addressed.

> Other than this, I dont see the log search configuration (grok patterns) as a part of this PR. How do we plan to update Storm's logsearch config when SAM is installed ?

The grok pattern should be part of Ambari, not SAM. I think just replacing default Storm logsearch config should be good to go.

File: conf/streamline_jaas.conf [hunk 1]
```diff
@@ -28,4 +28,13 @@ RegistryClient {
   storeKey=true
   useTicketCache=false
   principal="registy_client_principal@EXAMPLE.COM";
-};
\ No newline at end of file
+};
+Client {
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply conf/streamline_jaas.conf]: This is only the example for non-Ambari users, and it should be added to Ambari mpack as well.

File: streams/service/src/main/java/com/hortonworks/streamline/streams/service/TopologyCatalogResource.java [hunk 2]
```diff
@@ -56,10 +58,7 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.EnumSet;
-import java.util.Optional;
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply streams/service/src/main/java/com/hortonworks/streamline/streams/service/TopologyCatalogResource.java]: Addressed.

**Comment by arunmahadevan** [comment]
Raised followup issue https://github.com/hortonworks/streamline/issues/1071

**Comment by arunmahadevan** [comment]
+1

**Comment by arunmahadevan** [comment]
@HeartSaVioR thanks for the patch and working with the Ambari folks to sort out the issues. Merged to master.

**Pull request closed without merging**